### PR TITLE
Controlled buff wide inputs patch

### PR DESCRIPTION
**Pull request by seanpquinn**
Fix for #1725, fix > 1 bit width regression introduced by #1704

# Test plan

[test_controlled_buff_fix_wide_in.zip](https://github.com/logisim-evolution/logisim-evolution/files/11240647/test_controlled_buff_fix_wide_in.zip)

image

### CODE REVIEW

File: src/main/java/com/cburch/logisim/std/gates/ControlledBuffer.java [hunk 1]
```diff
@@ -247,10 +247,10 @@ public void propagate(InstanceState state) {
...
       final var in = state.getPortValue(1);
-      if (in == Value.TRUE || in == Value.FALSE) {
+      if (in.isFullyDefined()) {
         state.setPort(0, isInverter ? in.not() : in, GateAttributes.DELAY);
       } else if (in == Value.UNKNOWN) {
-        state.setPort(0, Value.createUnknown(width), GateAttributes.DELAY);
+        state.setPort(0, isInverter ? Value.createError(width) : Value.createUnknown(width), GateAttributes.DELAY);
       } else if (in == Value.ERROR) {
         state.setPort(0, Value.createError(width), GateAttributes.DELAY);
       }
```
Comment (davidhutchens) [inline src/main/java/com/cburch/logisim/std/gates/ControlledBuffer.java:256]: That still doesn't work. Value.UNKNOWN and Value.ERROR are one-bit values. They will not be equal to any value with more than one bit. isFullyDefined() means no errors or unknowns. Any multi-bit value may have some of each of the 4 values. Each bit should be processed independently. So no test on `in` makes sense.

All you need is the line:
`state.setPort(0, isInverter ? in.not() : in, GateAttributes.DELAY);`
That will pass through all the original values for the buffer and negate 0/1 and make U & E become E for the inverter.

File: snap/gui/logisim-evolution.desktop [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+Comment=Logisim-evolution is educational software for designing and simulating digital logic circuits.
+GenericName=Digital logic design tool and simulator
+Exec=logisim-evolution %f
+Icon=${SNAP}/meta/gui/logisim-icon-128.png
+Type=Application
+StartupNotify=false
+Categories=Electronics
+MimeType=application/x-logisim-circuit;
+Actions=new-empty-window;
+Keywords=circuits;simulation;logic;HDL;chronogram;
```
Comment (davidhutchens) [inline snap/gui/logisim-evolution.desktop:12]: A PR should be narrow in purpose. You shouldn't include a snap update in a controlled buffer patch PR.

**Review by davidhutchens — CHANGES_REQUESTED** [review]

File: snap/gui/logisim-evolution.desktop (hunk 2 above)
Comment (seanpquinn) [reply snap/gui/logisim-evolution.desktop:12]: Oops, this is bad VCS on my end: seems the snap PR got mixed up here, fixing

**Comment by seanpquinn** [comment]
Would unit tests for above cases be useful?

**Comment by davidhutchens** [comment]
@seanpquinn: At present, we lack jUnit tests for much of the core of the program. Efforts in that direction are welcome. All of our current tests pass and they do not need to be updated with this PR.

**Review by davidhutchens — APPROVED** [review]
This should fix the regression and adjust the controlled buffer output to pass U from `in` when selected. The controlled inverter will, when selected, pass E when given U as the not gate does.

**Pull request merged**